docs/rfc_service_layer_document.md

```diff
@@ -0,0 +1,206 @@
+                raise PermissionDenied("You do not have permission to disband this team.")
+
+        @conditional_swagger_auto_schema(
+            operation_id="cyberstorm.team.disband",
+            tags=["cyberstorm"],
+            responses={status.HTTP_204_NO_CONTENT: ""},
+        )
+        def delete(self, request, *args, **kwargs):
+            return super().delete(request, *args, **kwargs)
+    ```
+
+    The issue with this view is that it contains business logic related to deleting a team directly from the database. The view inherits from two other classes and performs permissions checking. The deletion logic and permissions checking could be moved elsewhere.
+
+## 2.2 Proposed Solution
+
+Move business logic from views to a separate service layer in each app with a clear domain. This will simplify views, making them easier to maintain, and will facilitate testing views and services in isolation.
+
+-   **Example of File Structure**:
+
+    ```
+    my_app/
+        ├── services/
+        │   ├── __init__.py
+        │   ├── team.py
+        ├── views/
+        │   ├── __init__.py
+        │   ├── team.py
+        └── tests/
+            ├── __init__.py
+            ├── test_services/
+            │   ├── __init__.py
+            │   ├── test_team.py
+            └── test_views/
+                ├── __init__.py
+                ├── test_team.py
+    ```
+
+# 3. Architecture Overview
+
+-   **Service Layer**: A new `services` module in each app will house files with functions containing core business logic (e.g., permissions checks, object manipulation). These functions will be small, focused on a single task, and will raise appropriate exceptions when necessary.
+
+-   **Views**: Views will be simplified to handle requests, responses, serialization, and the delegation of tasks.
+
+-   **Tests**: Tests will be updated to test the service layer in isolation, and views will be tested to ensure they handle requests and responses correctly.
+
+-   **Data Flow**:
+
+    1. The view receives a request.
+    2. The view delegates the business logic to the service layer.
+    3. The service layer performs the business logic, accesses the database, and returns a response.
+    4. The view handles the response and returns it to the client.
+
+    **Example**:
+
+    -   Request → View Layer (delegates) → Service Layer (accesses) → Database
+    -   Database → Service Layer (returns) → View Layer (returns) → Response
+
+# 4. Design Details
+
+## 4.1 Service Layer Design
+
+-   **Structure**:
+
+    -   Each app will have a new `services` directory.
+    -   Each service directory will contain files with functions that encapsulate core business logic (e.g., permissions checks, object manipulation).
+    -   Functions will be small, focused on a single task, and will raise appropriate exceptions when necessary.
+
+-   **Example**:
+
+    ```python
+    # file: services/team.py
+    def disband_team(team_name: str, user: User) -> None:
```

Comment:
I'd encourage us standardizing a specific term to indicate "the user that is executing the action" to avoid ambiguity when operating _on_ an user. In cases where these are the same, I would probably still separate them and simply feed in the same user in both variables.

`agent` has been used in a few niche locations so far and would be my suggestion, but another one I'm familiar with is  `actor`. Other suggestions welcome too